File: Java_version/app/src/beta/java/com/termux/wayland/beta/scanner/DesktopFileScanner.java
package com.termux.wayland.beta.scanner;

import com.termux.wayland.beta.data.LinuxApp;
import com.termux.wayland.beta.data.LinuxApp.AppSource;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Fallback : scanne les fichiers .desktop pour découvrir
 * des apps GUI non répertoriées dans apps.json.
 *
 * Note : les paquets Termux natifs n'ont PAS de .desktop,
 * ce scanner sert pour les apps depuis proot-distros ou sources tierces.
 *
 * Priorité DESKTOP = la plus basse dans le merge final.
 */
public class DesktopFileScanner {

    // Chemins à scanner
    private static final String[] SCAN_PATHS = {
        "/data/data/com.termux/files/usr/share/applications",
        "/data/data/com.termux/files/home/.local/share/applications"
    };

    // Singleton
    private static DesktopFileScanner instance;

    private DesktopFileScanner() {}

    public static synchronized DesktopFileScanner getInstance() {
        if (instance == null) {
            instance = new DesktopFileScanner();
        }
        return instance;
    }

    /**
     * Scanne tous les chemins connus et retourne les apps trouvées.
     */
    public List<LinuxApp> scanApps() {
        List<LinuxApp> apps = new ArrayList<>();

        for (String path : SCAN_PATHS) {
            File dir = new File(path);
            if (!dir.exists() || !dir.isDirectory()) continue;

            File[] files = dir.listFiles(f -> f.getName().endsWith(".desktop"));
            if (files == null) continue;

            for (File file : files) {
                LinuxApp app = parseDesktopFile(file);
                if (app != null) apps.add(app);
            }
        }

        return apps;
    }

    // ─── Parseur .desktop (format INI simplifié) ──────────────────────────────

    private LinuxApp parseDesktopFile(File file) {
        Map<String, String> props = new HashMap<>();

        try {
            BufferedReader reader = new BufferedReader(new FileReader(file));
            String line;
            while ((line = reader.readLine()) != null) {
                String trimmed = line.trim();
                // Ignorer commentaires et sections
                if (trimmed.startsWith("#") || trimmed.startsWith("[")) continue;
                int eq = trimmed.indexOf('=');
                if (eq > 0) {
                    String key   = trimmed.substring(0, eq).trim();
                    String value = trimmed.substring(eq + 1).trim();
                    // Garder seulement la première occurrence
                    if (!props.containsKey(key)) {
                        props.put(key, value);
                    }
                }
            }
            reader.close();
        } catch (Exception e) {
            return null;
        }

        // Filtres d'exclusion
        if ("true".equals(props.get("NoDisplay"))) return null;
        if (!"Application".equals(props.get("Type"))) return null;

        String name = props.get("Name");
        String execRaw = props.get("Exec");
        if (name == null || execRaw == null) return null;

        String exec = cleanExec(execRaw);
        if (exec.isEmpty()) return null;

        // Exclure les apps terminal-only sans icône
        boolean isTerminalApp = "true".equals(props.get("Terminal"));
        boolean hasIcon       = props.containsKey("Icon");
        if (isTerminalApp && !hasIcon) return null;

        // Id depuis le nom de fichier
        String id = file.getName()
                .replace(".desktop", "")
                .toLowerCase()
                .replace(" ", "-");

        String category = guessCategory(props.getOrDefault("Categories", ""));

        return new LinuxApp(
            id,
            name,
            id,     // packageName inconnu → on utilise l'id
            exec,
            category,
            props.get("Icon"),
            true,
            "unknown",
            props.getOrDefault("Comment", ""),
            AppSource.DESKTOP
        );
    }

    /**
     * Nettoie la valeur Exec= en retirant les placeholders %f %u %F %U etc.
     */
    private String cleanExec(String exec) {
        return exec.replaceAll("%[fFuUdDnNickvm]", "").trim();
    }

    /**
     * Déduit une catégorie depuis la chaîne Categories= du .desktop.
     */
    private String guessCategory(String categories) {
        String cats = categories.toLowerCase();
        if (cats.contains("webbrowser") || cats.contains("network")) return "web";
        if (cats.contains("graphics")   || cats.contains("photography")) return "graphics";
        if (cats.contains("development")|| cats.contains("ide"))         return "dev";
        if (cats.contains("office")     || cats.contains("wordprocessor"))return "office";
        if (cats.contains("video")      || cats.contains("audio") ||
            cats.contains("music")      || cats.contains("player"))       return "media";
        if (cats.contains("system")     || cats.contains("filemanager"))  return "system";
        if (cats.contains("utility")    || cats.contains("calculator"))   return "tools";
        return "tools";
    }
}

File: Kotlin_version/app/src/beta/java/com/termux/wayland/kotlin/beta/scanner/DesktopFileScanner.kt
package com.termux.wayland.kotlin.beta.scanner

import com.termux.wayland.kotlin.beta.data.AppSource
import com.termux.wayland.kotlin.beta.data.LinuxApp
import java.io.File

/**
 * Fallback : scanne les fichiers .desktop pour découvrir
 * des apps GUI non répertoriées dans apps.json.
 *
 * Note : les paquets Termux natifs n'ont PAS de .desktop,
 * donc ce scanner ne sert que pour les apps installées
 * depuis des proot-distros ou des sources tierces.
 *
 * Les résultats sont toujours de priorité DESKTOP (la plus basse).
 */
object DesktopFileScanner {

    // Chemins à scanner (ordre de priorité)
    private val SCAN_PATHS = listOf(
        "/data/data/com.termux/files/usr/share/applications",
        "/data/data/com.termux/files/home/.local/share/applications"
    )

    /**
     * Scanne tous les chemins connus et retourne les apps trouvées.
     * Les apps sans Exec= ou marquées NoDisplay=true sont ignorées.
     */
    fun scanApps(): List<LinuxApp> {
        val apps = mutableListOf<LinuxApp>()

        for (path in SCAN_PATHS) {
            val dir = File(path)
            if (!dir.exists() || !dir.isDirectory) continue

            dir.listFiles { f -> f.extension == "desktop" }
                ?.forEach { file ->
                    parseDesktopFile(file)?.let { apps.add(it) }
                }
        }

        return apps
    }

    // ─── Parseur .desktop (format INI simplifié) ──────────────────────────────

    private fun parseDesktopFile(file: File): LinuxApp? {
        val props = mutableMapOf<String, String>()

        try {
            file.bufferedReader().forEachLine { line ->
                val trimmed = line.trim()
                // Ignorer commentaires et sections
                if (trimmed.startsWith("#") || trimmed.startsWith("[")) return@forEachLine
                val eq = trimmed.indexOf('=')
                if (eq > 0) {
                    val key   = trimmed.substring(0, eq).trim()
                    val value = trimmed.substring(eq + 1).trim()
                    // On garde seulement la première occurrence (locale par défaut)
                    if (!props.containsKey(key)) {
                        props[key] = value
                    }
                }
            }
        } catch (e: Exception) {
            return null
        }

        // Filtres d'exclusion
        if (props["NoDisplay"] == "true") return null
        if (props["Type"] != "Application") return null

        val name = props["Name"] ?: return null
        val exec = props["Exec"]?.cleanExec() ?: return null

        // Détecter si l'app est graphique
        // On exclut les apps Terminal=true qui ne font qu'ouvrir un terminal texte
        // (sauf si elles ont une icône = elles ont probablement une vraie UI)
        val isTerminalApp = props["Terminal"] == "true"
        val hasIcon = props["Icon"] != null

        // Si Terminal=true et pas d'icône → probablement pas une vraie GUI
        if (isTerminalApp && !hasIcon) return null

        // Construire un id à partir du nom de fichier
        val id = file.nameWithoutExtension.lowercase().replace(" ", "-")

        // Déterminer la catégorie depuis Categories=
        val category = guessCategory(props["Categories"] ?: "")

        return LinuxApp(
            id             = id,
            name           = name,
            packageName    = id,   // on ne connaît pas le paquet exact
            exec           = exec,
            category       = category,
            icon           = props["Icon"],
            requiresDisplay= true,
            repo           = "unknown",
            description    = props["Comment"] ?: "",
            source         = AppSource.DESKTOP
        )
    }

    /**
     * Nettoie la valeur Exec= en retirant les placeholders %f %u %F %U etc.
     */
    private fun String.cleanExec(): String {
        return this.replace(Regex("%[fFuUdDnNickvm]"), "").trim()
    }

    /**
     * Déduit une catégorie depuis la chaîne Categories= du .desktop.
     */
    private fun guessCategory(categories: String): String {
        val cats = categories.lowercase()
        return when {
            "webbrowser" in cats || "network" in cats  -> "web"
            "graphics" in cats || "photography" in cats-> "graphics"
            "development" in cats || "ide" in cats     -> "dev"
            "office" in cats || "wordprocessor" in cats-> "office"
            "video" in cats || "audio" in cats ||
                "music" in cats || "player" in cats    -> "media"
            "system" in cats || "filemanager" in cats  -> "system"
            "utility" in cats || "calculator" in cats  -> "tools"
            else                                        -> "tools"
        }
    }
}
